rfcs/2021-05-23-groups2.md

```diff
@@ -0,0 +1,134 @@
+
+Two approaches are possible:
+
+1. Each agent on a regular interval sends to all agents the sender IDs and the digests of the last messages from all agents they communicate with.
+To avoid `(n-1)*n` messages for each group integrity verification, IDs and digests from all senders can be compacted in one messages:
+
+```
+broadcast: %s"CHECK" SP memId1 ":" msgId1 ":" hash1 ":" status1 SP memId2 ":" msgId2 ":" hash2 ":" status2
+```
+
+A side question is that we currently do not support large agent messages; possibly we should support messages larger than SMP block size to simplify this and other scenarios, similarly to how websockets protocol does it. There might still be a limit to how large the full message can be. That probably requires re-thinking of how messages are managed and separation of message reception from the servers and message delivery to the clients, but this is likely to be required anyway for when we start running the agent in the background.
+
+There may be two situations to consider:
+
+1) The recipient of this verification message can have the same message as the last one or they can have more messages received - both such scenarios are ok and do not necessarily indicate a problem, as they just might be slightly ahead in receiving the messages. They would make the last matching message as ok, and the latter messages would remain unknown - it does not indicate the lack of integrity of the group, although if the next integrity check without the same messages does (this needs to be clarified).
+2) The recipient of this verification message can have fewer messages received. This situation can be resolved in several ways:
+- try to retrieve all messages from the queue that is behind. It might work, but it may be that the sender is simply trying to send messages because the network is down.
+- wait until the next integrity check and only report integrity violation if during the next integrity check they still cannot reconcile the previous integrity check. This is probably an acceptable compromise.
+
+2. Each agent sends message receipts to sending agents - message receipts would contain a signature of the message hashes. In the agent uses the same verification key for each member of the group, as considered below, these receipts can be re-broadcasted to other agents (again, grouping them in one message to avoid `(n-1)*n` messages) as a proof that the messages were delivered.
+
+Comparing with the first approach, there are pros and cons:
+- pros:
+  - the sender would only send such integrity check messages when they have send a message to all parties, thus avoiding the situation when some messages might have been not yet sent (or failed to send and they are retrying).
+- cons:
+  - integrity check message mush contain a signature per member, so it would be substantially larger.
+  - agents must use the same verification key for all members in the group, complicating the group and connections management.
```

Comment:
Why does it complicate managment?

Reply:
because it doesn't allow treating connections as black boxes that generate their own keys, it requires breaking the abstraction boundary and passing the same key from the outside of connection.

It effectively creates a long-term group-wide form of identity which is something that may need to be rotated in some scenarios and in general complicates things.

We already have a shared id per member, adding a shared key adds complexity.